src/mongocrypt-ctx-decrypt.c

```diff
@@ -39,68 +59,92 @@ _replace_FLE2IndexedEncryptedValue_with_plaintext (void *ctx,
    BSON_ASSERT_PARAM (in);
    BSON_ASSERT_PARAM (out);
 
-   if (!mc_FLE2IndexedEncryptedValue_parse (iev, in, status)) {
-      goto fail;
-   }
-
+   CHECK_AND_RETURN (mc_FLE2IndexedEncryptedValue_parse (iev, in, status));
    const _mongocrypt_buffer_t *S_KeyId =
       mc_FLE2IndexedEncryptedValue_get_S_KeyId (iev, status);
-   if (!S_KeyId) {
-      goto fail;
-   }
+   CHECK_AND_RETURN (S_KeyId);
+   CHECK_AND_RETURN_KB_STATUS (
+      _mongocrypt_key_broker_decrypted_key_by_id (kb, S_KeyId, &S_Key));
+   CHECK_AND_RETURN (mc_FLE2IndexedEncryptedValue_add_S_Key (
+      kb->crypt->crypto, iev, &S_Key, status));
+   const _mongocrypt_buffer_t *K_KeyId =
+      mc_FLE2IndexedEncryptedValue_get_K_KeyId (iev, status);
+   CHECK_AND_RETURN (K_KeyId);
+   CHECK_AND_RETURN_KB_STATUS (
+      _mongocrypt_key_broker_decrypted_key_by_id (kb, K_KeyId, &K_Key));
+   CHECK_AND_RETURN (mc_FLE2IndexedEqualityEncryptedValue_add_K_Key (
```

Comment:
Not related to these changes. Suggest renaming `mc_FLE2IndexedEqualityEncryptedValue_add_K_Key` to `mc_FLE2IndexedEncryptedValue_add_K_Key` since it is used for both Equality and Range payloads.

Reply:
I did have the same thought, actually.  Happy to include it.